beard.js

```diff
@@ -1,210 +1,207 @@
-    if:         (/^if\s+([^]*)$/),
-    elseIf:     (/^else\s+if\s+([^]*)$/),
-    else:       (/^else$/),
-    for:        (/^for\s+([$A-Za-z_][0-9A-Za-z_]*)(?:\s*,\s*([$A-Za-z_][0-9A-Za-z_]*))?\s+in\s+(.*)$/),
-    each:       (/^each\s+([$A-Za-z_][0-9A-Za-z_]*)(?:\s*,\s*([$A-Za-z_][0-9A-Za-z_]*))?\s+in\s(.*)$/),
-    end:        (/^end$/)
-  };
-
-  const parse = {
-    include:    (_, includePath) => `_capture(compiled("${includePath}", path)(_context));`,
-    includeFn:  (_, __, includePath, data) => `_context.locals.push({${data}}); _capture(compiled("${includePath}", path)(_context)); _context.locals.pop();`,
-    block:      (_, blockname) => `_blockName = "${blockname}"; _blockCapture = "";`,
-    blockEnd:   () => 'eval(`var ${_blockName} = _blockCapture`); _context.globals[_blockName] = _blockCapture; _blockName = null;',
-    encode:     (_, statement) => `_encode(${statement});`,
-    comment:    () => '',
-    if:         (_, statement) => `if (${statement}) {`,
-    elseIf:     (_, statement) => `} else if (${statement}) {`,
-    else:       () => '} else {',
-    end:        () => '}',
-    for: (_, key, value, object) => {
-      if (!value) key = (value = key, 'iterator' + iterator++);
-      return `for (var ${key} in ${object}){ var ${value} = ${object}[${key}];`;
-    },
-    each: (_, iter, value, array) => {
-      if (!value) iter = (value = iter, 'iterator' + iterator++);
-      const length = 'length' + iterator++;
-      return `for (var ${iter} = 0, ${length} = ${array}.length; ${iter} < ${length}; ${iter}++) { var ${value} = ${array}[${(iter)}];`;
-    }
-  };
-
-  function parser(match, inner) {
-    const prev = inner;
-    inner = inner
-      .replace(exps.include, parse.include)
-      .replace(exps.includeFn, parse.includeFn)
-      .replace(exps.block, parse.block)
-      .replace(exps.blockEnd, parse.blockEnd)
-      .replace(exps.encode, parse.encode)
-      .replace(exps.comment, parse.comment)
-      .replace(exps.end, parse.end)
-      .replace(exps.else, parse.else)
-      .replace(exps.elseIf, parse.elseIf)
-      .replace(exps.if, parse.if)
-      .replace(exps.each, parse.each)
-      .replace(exps.for, parse.for);
-
-    const middle = inner === prev && !/^:/.test(inner)
-      ? `_capture(${inner});`
-      : inner.replace(/\t|\n|\r|^:/, '');
-
-    return `"); ${middle} _capture("`;
+  asset(p, path) {
+    const absolutePath = resolvePath(p, path);
+    return this.opts.asset(absolutePath) || absolutePath;
   }
 
-  function compiled(path, parentPath) {
-    const fullPath = resolvePath(path, parentPath);
-    if (opts.cache) {
-      const str = opts.templates[fullPath];
-      const key = hash(fullPath);
-      if (!fnCache[key]) fnCache[key] = compile(str, fullPath);
-      return fnCache[key];
-    } else {
-      const str = fs.readFileSync(pathMap[fullPath], 'utf8');
-      return compile(str, fullPath);
+  render(path, data = {}) {
+    const context = {
```

Comment:
@joeosburn doesn't `context` get mutated throughout the lifecycle of a render? In which case this should be `let` instead of `const`